### ISSUE TITLE
Significant memory leak

### ISSUE THREAD

**Issue by phet-steele** [phetsims/charges-and-fields#83]
Heap collected every two minutes for just 10 minutes using `?fuzzMouse=100`. Numbers collected on OSX 10.11.6 Chrome; iPad 2 Safari crashed after 10 minutes.

| MB | minute |
| --- | --- |
| 17.5 | 0 |
| 74.5 | 2 |
| 122 | 4 |
| 169 | 6 |
| 216 | 8 |
| 270 | 10 |

For phetsims/tasks#661.
Troubleshooting information:
Name: ‪Charges And Fields‬
URL: http://www.colorado.edu/physics/phet/dev/html/charges-and-fields/1.0.0-dev.15/charges-and-fields_en.html
Version: 1.0.0-dev.15 2016-07-28 14:16:27 UTC
Features missing: touch
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.106 Safari/537.36
Language: en-US
Window: 1920x1002
Pixel Ratio: 2/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 32 uniform: 1024
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 16)
Max viewport: 16384x16384
OES_texture_float: true
Dependencies JSON: {"assert":{"sha":"7d27130a","branch":"master"},"axon":{"sha":"56e95e01","branch":"master"},"babel":{"sha":"b0a15003","branch":"master"},"brand":{"sha":"f0b1f7da","branch":"master"},"charges-and-fields":{"sha":"9163b74d","branch":"master"},"chipper":{"sha":"bdc6b525","branch":"master"},"dot":{"sha":"6ee3c51e","branch":"master"},"joist":{"sha":"50af4ae6","branch":"master"},"kite":{"sha":"88f0c3a0","branch":"master"},"phet-core":{"sha":"c48bf320","branch":"master"},"phetcommon":{"sha":"83ea84c8","branch":"master"},"scenery":{"sha":"942889ad","branch":"master"},"scenery-phet":{"sha":"eaa4be69","branch":"master"},"sherpa":{"sha":"5ddfccd4","branch":"master"},"sun":{"sha":"85b7fc95","branch":"master"},"tandem":{"sha":"6cc3e843","branch":"master"}}

**Comment by andrewadare** [comment]
This is a list of `link` calls (singly indented) and `unlink` calls (doubly indented) in the sim:

```
/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/model/ChargesAndFieldsModel.js:
    addedChargedParticle.isUserControlledProperty.link( userControlledListener );
    addedChargedParticle.positionProperty.link( positionListener );
    addedElectricFieldSensor.positionProperty.link( positionListener );
    addedElectricFieldSensor.isUserControlledProperty.link( userControlledListener );
        addedChargedParticle.isUserControlledProperty.unlink( userControlledListener );
        addedChargedParticle.isActiveProperty.unlink( isActiveListener );
        addedChargedParticle.positionProperty.unlink( positionListener );
        addedElectricFieldSensor.isUserControlledProperty.unlink( userControlledListener );
        addedElectricFieldSensor.positionProperty.unlink( positionListener );

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ChargedParticleNode.js:
    chargedParticle.positionProperty.link( positionListener );
    availableModelBoundsProperty.link( availableModelBoundsPropertyListener );
    chargedParticle.isInteractiveProperty.link( function() {
        availableModelBoundsProperty.unlink( availableModelBoundsPropertyListener );
        chargedParticle.positionProperty.unlink( positionListener );

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricFieldCanvasNode.js:
    isVisibleProperty.link( invalidateSelfListener ); // visibility change
    isElectricFieldDirectionOnlyProperty.link( invalidateSelfListener ); // visibility change
    particle.positionProperty.link( invalidateSelfListener );
    isVisibleProperty.linkAttribute( this, 'visible' );
        particle.positionProperty.unlink( invalidateSelfListener );

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricFieldSensorNode.js:
    electricFieldSensor.electricFieldProperty.link( electricFieldListener );
    electricFieldSensor.isActiveProperty.link( function( isActive ) {
    areValuesVisibleProperty.link( areValuesVisibleListener );
    isDirectionLabelVisibleProperty.link( isDirectionLabelVisibleListener );
    electricFieldSensor.positionProperty.link( positionListener );
    electricFieldSensor.isInteractiveProperty.link( function() {
    availableModelBoundsProperty.link( availableModelBoundsPropertyListener );
        electricFieldSensor.positionProperty.unlink( positionListener );
        electricFieldSensor.electricFieldProperty.unlink( electricFieldListener );
        areValuesVisibleProperty.unlink( areValuesVisibleListener );
        isDirectionLabelVisibleProperty.unlink( isDirectionLabelVisibleListener );
        availableModelBoundsProperty.unlink( availableModelBoundsPropertyListener );

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricPotentialCanvasNode.js:
    isVisibleProperty.link( invalidateSelfListener ); // visibility change
    particle.positionProperty.link( invalidateSelfListener );
    isVisibleProperty.linkAttribute( this, 'visible' );
        particle.positionProperty.unlink( invalidateSelfListener );

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricPotentialLinesNode.js:
    locationProperty.link( locationFunction );
    areValuesVisibleProperty.linkAttribute( labelsNode, 'visible' );
        locationProperty.unlink( locationFunction);

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricPotentialMobileWebGLNode.js:
    isVisibleProperty.link( invalidateSelfListener ); // visibility change
    particle.positionProperty.link( invalidateSelfListener );
        particle.positionProperty.unlink( invalidateSelfListener );

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricPotentialWebGLNode.js:
    isVisibleProperty.link( invalidateSelfListener ); // visibility change
    particle.positionProperty.link( invalidateSelfListener );
        particle.positionProperty.unlink( invalidateSelfListener );
```

**Comment by andrewadare** [comment]
From heap snapshot analysis, it looks like ArrowNode might be part of the problem:
![image](https://cloud.githubusercontent.com/assets/154163/17487323/97d1ca1c-5d51-11e6-81bb-790397ff5fa5.png)

**Comment by andrewadare** [comment]
Following up from the previous entry, here's a simpler representation of the properties that are linked and unlinked by file. Several missing `unlink`s are highlighted with a `-->`.

```
/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/model/ChargesAndFieldsModel.js:
    addedChargedParticle.isUserControlledProperty
    addedElectricFieldSensor.positionProperty
    addedChargedParticle.positionProperty
    addedElectricFieldSensor.isUserControlledProperty
        addedElectricFieldSensor.isUserControlledProperty
        addedChargedParticle.isUserControlledProperty
        addedElectricFieldSensor.positionProperty
        addedChargedParticle.positionProperty
        addedChargedParticle.isActiveProperty

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ChargedParticleNode.js:
    chargedParticle.positionProperty
    availableModelBoundsProperty
    chargedParticle.isInteractiveProperty
        chargedParticle.positionProperty
        availableModelBoundsProperty
    --> chargedParticle.isInteractiveProperty

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricFieldCanvasNode.js:
    particle.positionProperty
    isVisibleProperty
    isElectricFieldDirectionOnlyProperty
        particle.positionProperty
    --> isVisibleProperty
    --> isElectricFieldDirectionOnlyProperty

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricFieldSensorNode.js:
    electricFieldSensor.isActiveProperty
    isDirectionLabelVisibleProperty
    availableModelBoundsProperty
    areValuesVisibleProperty
    electricFieldSensor.positionProperty
    electricFieldSensor.isInteractiveProperty
    ChargesAndFieldsColors
    electricFieldSensor.electricFieldProperty
        electricFieldSensor.isActiveProperty
        isDirectionLabelVisibleProperty
        availableModelBoundsProperty
        areValuesVisibleProperty
        electricFieldSensor.isInteractiveProperty
        electricFieldSensor.positionProperty
        ChargesAndFieldsColors
        electricFieldSensor.electricFieldProperty

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricPotentialCanvasNode.js:
    particle.positionProperty
    isVisibleProperty
        particle.positionProperty
    --> isVisibleProperty

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricPotentialLinesNode.js:
    locationProperty
    areValuesVisibleProperty
    ChargesAndFieldsColors
        locationProperty
        ChargesAndFieldsColors
    --> areValuesVisibleProperty

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricPotentialMobileWebGLNode.js:
    particle.positionProperty
    isVisibleProperty
        particle.positionProperty
    --> isVisibleProperty

/Users/adare/phetsims/charges-and-fields/js/charges-and-fields/view/ElectricPotentialWebGLNode.js:
    particle.positionProperty
    isVisibleProperty
        particle.positionProperty
    --> isVisibleProperty
```

**Comment by andrewadare** [comment]
I have added dispose() methods where I was able to find them missing. Heap snapshot comparisons do not indicate that the problem is fully resolved. There are many mentions of generic objects (Object, closue, Array) and scenery objects, but I am having a very difficult time tracing their origin to the sim.

**Comment by andrewadare** [comment]
@jonathanolson If you get a chance, I'm wondering if you could take a look at window.addEventListener in ChargesAndFieldsColors. Is it possible this function is contributing to this leak?

In heap snapshot comparisons, arrays are the top offender, when sorted by '# delta' or 'size delta'. (I'm not sure about the difference between (array) vs. Array). When I drill further, I see entries like this:
![image](https://cloud.githubusercontent.com/assets/154163/17713697/64bc31e4-63b9-11e6-8367-078814a7b6c5.png)

In other words, "Window" is mentioned a lot. In any case, I've been having quite a tough time tracing these console messages back to places in the sim that need attention, so any pointers would be greatly appreciated.

**Comment by andrewadare** [comment]
@jonathanolson never mind, I was misguided in my hypothesis. Sorry for the noise.

**Comment by phet-steele** [comment]
This sim is all sealed up, nice work!
